Mathlib/Topology/IsClosedRestrict.lean

```diff
@@ -0,0 +1,140 @@
+on `Sᶜ` and are arbitrary on `S`.
+`reorderRestrictProd` sends a term of that type to `Π j, α j` by looking for the value at `j`
+in one part of the product or the other depending on whether `j` is in `S` or not. -/
+noncomputable def reorderRestrictProd (S : Set ι) (s : Set (Π j, α j))
+    (p : Sᶜ.restrict '' s × (Π i : S, α i)) :
+    Π j, α j :=
+  fun j ↦ if h : j ∈ S
+    then (p.2 : Π j : ↑(S : Set ι), α j) ⟨j, h⟩
+    else (p.1 : Π j : ↑(Sᶜ : Set ι), α j) ⟨j, h⟩
+
+@[simp]
+lemma reorderRestrictProd_of_mem (p : Sᶜ.restrict '' s × (Π i : S, α i)) (j : S) :
+    reorderRestrictProd S s p j = (p.2 : Π j : ↑(S : Set ι), α j) j := by
+  have hj : ↑j ∈ S := j.prop
+  simp [reorderRestrictProd, hj]
+
+@[simp]
+lemma reorderRestrictProd_of_compl (p : Sᶜ.restrict '' s × (Π i : S, α i)) (j : (Sᶜ : Set ι)) :
+    reorderRestrictProd S s p j = (p.1 : Π j : ↑(Sᶜ : Set ι), α j) j := by
+  have hj : ↑j ∉ S := j.prop
+  simp [reorderRestrictProd, hj]
+
+@[simp]
+lemma restrict_compl_reorderRestrictProd (p : Sᶜ.restrict '' s × (Π i : S, α i)) :
+    Sᶜ.restrict (reorderRestrictProd S s p) = p.1 := by ext; simp
+
+lemma continuous_reorderRestrictProd [∀ i, TopologicalSpace (α i)] :
+    Continuous (reorderRestrictProd S s) := by
+  refine continuous_pi fun j ↦ ?_
+  simp only [reorderRestrictProd]
+  split_ifs with h
+  · fun_prop
+  · exact ((continuous_apply _).comp continuous_subtype_val).comp continuous_fst
+
+lemma reorderRestrictProd_mem_preimage_image_restrict (p : Sᶜ.restrict '' s × (Π i : S, α i)) :
+    reorderRestrictProd S s p ∈ Sᶜ.restrict ⁻¹' (Sᶜ.restrict '' s) := by
+  obtain ⟨y, hy_mem_s, hy_eq⟩ := p.1.2
+  exact ⟨y, hy_mem_s, hy_eq.trans (restrict_compl_reorderRestrictProd p).symm⟩
+
+@[simp]
+lemma reorderRestrictProd_restrict_compl (x : Sᶜ.restrict ⁻¹' (Sᶜ.restrict '' s)) :
+    reorderRestrictProd S s ⟨⟨Sᶜ.restrict x, x.2⟩, fun i ↦ (x : Π j, α j) i⟩ = (x : Π j, α j) := by
+  ext; simp [reorderRestrictProd]
+
+/-- Homeomorphism between the set of functions that concide with a given set of functions away
+from a given set `S`, and dependent functions away from `S` times any value on `S`. -/
+noncomputable
+def preimageImageRestrict (α : ι → Type*) [∀ i, TopologicalSpace (α i)]
+    (S : Set ι) (s : Set (Π j, α j)) :
+    Sᶜ.restrict ⁻¹' (Sᶜ.restrict '' s) ≃ₜ Sᶜ.restrict '' s × (Π i : S, α i) where
+  toFun x := ⟨⟨Sᶜ.restrict x, x.2⟩, fun i ↦ (x : Π j, α j) i⟩
+  invFun p := ⟨reorderRestrictProd S s p, reorderRestrictProd_mem_preimage_image_restrict p⟩
+  left_inv x := by ext; simp
+  right_inv p := by ext <;> simp
+  continuous_toFun := by
+    refine Continuous.prod_mk ?_ ?_
+    · exact ((Pi.continuous_restrict _).comp continuous_subtype_val).subtype_mk _
+    · rw [continuous_pi_iff]
+      exact fun _ ↦ (continuous_apply _).comp continuous_subtype_val
+  continuous_invFun := continuous_reorderRestrictProd.subtype_mk _
+
+lemma image_snd_preimageImageRestrict [∀ i, TopologicalSpace (α i)] :
```

Comment:
This lemma might we worth a docstring (notably mentioning `IsCompact.isClosed_image_restrict`)